### ISSUE TITLE
Selection should never be unset

### ISSUE THREAD

**Issue by SamyPesse** [ianstormtaylor/slate#388]
@ianstormtaylor A few months ago, you opened [#36 `get rid of the concept of an "unset" selection, normalizing instead`](https://github.com/ianstormtaylor/slate/issues/36), but since then, it has been rollbacked by the commit [`fix selection setting for history`](https://github.com/ianstormtaylor/slate/commit/521f0d9f1d7d820389c8681c96429c02b35eb98a).

This is causing issues in a lot of places, and requires to maintain conditions like `if (!selection.startKey) { ...` everywhere. For example @bkniffler opened a PR to handle these cases for some methods: [#329](https://github.com/ianstormtaylor/slate/pull/329).

It's causing bugs in our application with some weird scenarios where the editor is focused but the selection is unset, and pressing enter or backspace throw critical errors.

Our current solution will be to add conditions like `if (!selection.startKey) return` everytime we have to work with the selection, even in `onKeyDown` where it seems illogical to get a keyboard event without a valid selection (we can reproduced this bug).

I think the selection can be focused or blurred; but **should never** be unset, and we should always normalize to keep a valid selection.

What are the reasons for having potential unset selections ?  
Should we go with adding conditions to handle this case and merge #329 ?

**Comment by Soreine** [comment]
Having to handle an unset selection requires more effort, for no apparent benefits. I think we should go back to trying to always normalize the selection.

We might want to avoid going back to calling the generic `selection.normalize` everywhere. Instead we could update the selection as each case fits. I think we mostly need two things:

- A function to reset the selection to the first valid place (beginning of document). Used as last resort.
- A function to update the selection to the nearest valid place that is not in a given node or range. I can only see the selection to become invalid when we remove nodes/text where the selection is. That's where it shines to have a transform such as this one.

**Comment by ianstormtaylor** [comment]
Thanks for opening this, down to discuss. The reason I actually added back in the `unsetSelection` transform and applied it in all of those cases was for proper history handling. 

Previously there were no "unset" selections, and in the case of performing a transform like `delete()`, the nodes would be deleted, and then the selection would be updated. However, when trying to reverse those changes, when performing an "undo", it becomes impossible. Because the first operation to be reversed will be `setSelection`, which will try to set the selection to a node that was deleted, since the reverse of `removeNode` (which is `insertNode`) won't have been applied yet.

Does that make sense?

What I did to solve that issue is to ensure that the selection is in an "unset" state while transforms that remove content are applied, so that it is always guaranteed to work properly in reverse for history.

I might have overlooked a simpler solution though. Given that constraint, what would you two propose we do?

**Comment by ianstormtaylor** [comment]
Separately, what are the places where you're bring forced to check for `!startKey`? If we can get a good sense of those, there are probably places where we can normalize that aren't those specific transforms.

**Comment by SamyPesse** [comment]
@ianstormtaylor I was able to make history work in #389, the main issues was the use of paths instead of keys when creating the `set_selection` operation, are you using paths instead of keys so that operations can work on similar document with different node keys across network (real-time collaboration) ?

Resolving the paths from the key failed because the node with the key doesn't exist after a `delete` operation, I think unsetting the selection looks like a hack, one solution might be to pass the previous document to the operation builder so that it can resolve the keys into paths using the right document instance.

> Separately, what are the places where you're bring forced to check for !startKey? If we can get a good sense of those, there are probably places where we can normalize that aren't those specific transforms.

For example we'll need this condition at [`examples/auto-markdown/index.js#L186`](https://github.com/ianstormtaylor/slate/blob/master/examples/auto-markdown/index.js#L186), this code can fail if you have for exampels tables with button controls to remove a row, clicking the button will remove the row, but if you press `Backspace` after clicking the button, the editor is focused but the selection is unset, so it fail with a critical error.

----

I'll keep working on https://github.com/ianstormtaylor/slate/pull/389 to prevent using keys instead of paths.

**Comment by Soreine** [comment]
> Previously there were no "unset" selections, and in the case of performing a transform like `delete()`, the nodes would be deleted, and then the selection would be updated.

That's the root of the problem. The `delete()` operation was going through an invalid state where the selection pointed to a non-existing node. Would it be possible instead to update the selection first, then delete the nodes? Ideally, we should be able to make every transforms, even the minimal ones, a series of transforms from valid states to valid states.

> Separately, what are the places where you're bring forced to check for !startKey?

Yes, from the moment we know some transforms can put the editor in a state with an unset selection, every piece of code that need to use the current selection must do the check (another example is the display of Bold and Italic buttons as active or not). Or we assume the unset state is an invalid state that should stay internal. Hence every user land transforms must make sure to update the selection accordingly after deleting nodes.

**Comment by ianstormtaylor** [comment]
@SamyPesse yup the paths are needed for collaboration as envisioned, to be able to work on a different document with different keys. Honestly I haven't gotten that far, and maybe the constraint of having the same keys could exist, not sure. But I think we'd have the same problem regardless, since when inverting them there would be a point where the selection cannot be applied, since the node doesn't exist.

@Soreine updating the selection first as currently written doesn't quite work, because there are cases where the node that will end up being selected doesn't exist yet. We could update it to be `0,0` so that it's not "invalid" but then that feels like a hack. Technically the transforms could be refactored in such a way that this is possible, but with the current distinction between `deleteAtRange()` and `delete()` where only `delete` cares about updating the selection makes that not possible—I think.

I'm not sure that unsetting the selection is a hack, it actually feels semantic?

What I'm curious about though is why the unset selections are "leaking" out of the transforms. (I'm probably missing something obvious here.) In the case of `delete()`, it calls `unsetSelection()` when it starts and `setSelection()` before it ends, so when would you run into unset selections occurring because of that? As long as the unsetting is always cleaned up before the transform has finished, won't that never be exposed? Eg.:

```js
return transform
  .unsetSelection()
  .deleteAtRange(selection)
  .moveTo(after)
```

Still not convinced unsetting is correct, but want to push back more 😄

**Comment by Soreine** [comment]
I see

> As long as the unsetting is always cleaned up before the transform has finished, won't that never be exposed

Yes, then slate's core and plugins need to be careful with that in their transforms. But it's cumbersome to keep the selection valid. Last time I encountered this problem was for unwrapping inlines. After unwrapping the inline, we merge together adjacent text nodes using `normalize`. So we need to update the selection and compute the new key and offset within the merged text node. It's not obvious. We should make it easier. Do you think we can provide some relevant selection normalization functions [as I said earlier](https://github.com/ianstormtaylor/slate/issues/388#issuecomment-254069228)?

**Comment by ianstormtaylor** [comment]
Ah! Yes, that makes sense that `Node.normalize` was forgetting it.

I think with @SamyPesse's work on getting normalization to all take place via the schema, we should actually use those same `normalize` schema functions to normalize the selection at the same time, for the specific validation rule case. And then ideally we could get rid of the need to be calling `Selection.normalize` in lots of (if not all) places.

**Comment by ianstormtaylor** [comment]
Closing since this will be solved in the #392 PR I believe.
